### PR TITLE
Instrument connector SPI with JFR

### PR DESCRIPTION
**Pull request by wendigo**
Adds the delegating Connector SPI implementations that are generating JFR events on method calls.

Screenshot 2023-03-12 at 21 06 06

### CODE REVIEW

**Comment by electrum** [comment]
This is interesting.

If we're going to check in the generated code, then we should clean it up and have it match the code style, so that we don't need checkstyle suppressions. Otherwise, we should generate it at build time rather than checking it in.

**Comment by wendigo** [comment]
Agree @electrum. This was just a raw idea that I had some time ago and discussed it with @mattstep during his visit in Warsaw, hence the WiP PR

**Comment by mattstep** [comment]
Yeah @wendigo lets chat when you're back from vacation, or in Boston, but I think we can associate relevant traces from opentelemetry using the trace/span id with these JFR events and even query id, that way we can connect the individual jfr events between nodes together and trace back to the specific inter-cluster api calls that resulted in what jfr events. It would give us ordering and parentage of events across the whole cluster.

**Comment by wendigo** [comment]
@hashhar @kokosing can you run this with secrets?

**Comment by kokosing** [comment]
/test-with-secrets sha=6a9ca5d191dee61374f7fb568d057e042960725d

**Comment by wendigo** [comment]
Failure in CI unrelated to this change: BigQuery FTE leftover tables cc @hashhar

File: plugin/trino-raptor-legacy/src/main/java/io/trino/plugin/raptor/legacy/RaptorConnectorFactory.java [hunk 1]
```diff
@@ -66,29 +69,31 @@ public Connector create(String catalogName, Map<String, String> config, Connecto
...
-                    binder.bind(PageSorter.class).toInstance(context.getPageSorter());
-                    binder.bind(TypeManager.class).toInstance(context.getTypeManager());
-                    binder.bind(CatalogName.class).toInstance(new CatalogName(catalogName));
-                },
-                metadataModule,
-                new BackupModule(backupProviders),
-                new StorageModule(),
-                new RaptorModule(),
-                new RaptorSecurityModule());
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(classLoader)) {
```
Comment (electrum) [inline plugin/trino-raptor-legacy/src/main/java/io/trino/plugin/raptor/legacy/RaptorConnectorFactory.java:72]: No need to pass in the class loader. We can simply do `getClass().getClassLoader()` here.

However, I don't understand why this change is needed at all, since `ConnectorFactory.create()` is already called with the correct class loader: https://github.com/trinodb/trino/blob/1ca635e31eb2215e064e054f2aef38053b8e5426/core/trino-main/src/main/java/io/trino/connector/DefaultCatalogFactory.java#L202-L204

Comment (electrum) [reply plugin/trino-raptor-legacy/src/main/java/io/trino/plugin/raptor/legacy/RaptorConnectorFactory.java:72]: I see the problem now. We should fix this in the instrumentation to preserve the correct class loader. We provide this undocumented guarantee for other plugin services. This type of behavior change is likely to break other connectors and is difficult for people to debug.

**Review by skrzypo987 — COMMENTED** [review]
I don't see a kill switch.
While that way of instrumentation has small overhead, it may still cause regression.
It is reasonable to expect queries with 10k+ access control checks. All of them are carried out single-threaded on a coordinator.

File: plugin/trino-raptor-legacy/src/main/java/io/trino/plugin/raptor/legacy/RaptorConnectorFactory.java (hunk 1 above)
Comment (wendigo) [reply plugin/trino-raptor-legacy/src/main/java/io/trino/plugin/raptor/legacy/RaptorConnectorFactory.java:72]: Fixed.
